hathor/dag_builder/tokenizer.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2024 Hathor Labs
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+from enum import Enum, auto
+from typing import Any, Iterator
+
+"""
+A domain specific language to describe DAGs.
+
+Syntax:
+
+    blockchain genesis a[2..5]  # create blocks a2, a3, a4, and a5 where a2's parent is the genesis block
+    blockchain pi a[5..7]       # create blocks a5, a6, and a7 where a5's parent is pi
+    a <-- b <-- c               # a is a parent of b which is a parent of c
+    a --> b --> c               # c is a parent of b which is a parent of a
+    a.out[i] <<< b              # b spends the i-th output of a
+    a < b < c                   # a must be created before b and b must be created before c
+    a > b > c                   # a must be created after b and b must be creater after c
+    a.attr = value              # set value of attribute attr to a
+
+Special attributes:
+    a.out[i] = 100 HTR          # set that the i-th output of a holds 100 HTR
+    a.out[i] = 100 TOKEN        # set that the i-th output of a holds 100 TOKEN where TOKEN is a custom token
+
+
+Example:
+
+    blockchain genesis a[0..300]
+    blockchain a300 b[0..20]
+    blockchain b4 c[0..10]
+
+    # reward lock
+    a300 < dummy
+
+    b11 --> tx1
+    b11 --> tx2
+
+    b14 --> tx1
+    b14 --> tx3
+
+    c3 --> tx1
+    c3 --> tx2
+
+    tx1 <-- tx2 <-- tx3
+
+    tx3 --> tx5 --> tx6
+
+    tx1.out[0] <<< tx2 tx3
+    tx1.out[0] <<< tx4
+
+    a0.out[0] <<< tx1
+
+    tx1.out[0] = 100 HTR [wallet1]
+    tx1.out[1] = 50 TK1  [wallet2]
+    tx2.out[0] = 75 USDC [wallet1]
+
+    USDC.out[0] = 100000 HTR
+
+    b5 < c0 < c10 < b20
+    b6 < tx3
+    b16 < tx4
+"""
+
+
+class TokenType(Enum):
+    BLOCKCHAIN = auto()
+    ATTRIBUTE = auto()
+    PARENT = auto()
+    SPEND = auto()
+    OUTPUT = auto()
+    ORDER_BEFORE = auto()
+
+
+Token = tuple[TokenType, tuple[Any, ...]]
```

Comment:
It would be better if you had strong types for each token constructor. But I'm fine if you don't want to refactor this now.

Reply:
I agree. But I feel that the syntax is so simple that it would be overkill for now.